### ISSUE TITLE
jnp.nanmean can overflow in the denominator for large arguments

### ISSUE THREAD

**Issue by andorardo** [google/jax#13033]
### Description

The following reproduces a phenomenon where the mean of a positive tensor can be negative:
```python
logits = np.random.uniform(size=[2048,64, 32000], low=-4, high=4)
jnp.nanmean(jnp.abs(logits))
=> DeviceArray(-83.333984, dtype=float32)
```

The reason is that the mean normalizer is summed using `np.int32` here: https://github.com/google/jax/blob/540835f979fe13a968f69b894dbafc981f122b36/jax/_src/numpy/reductions.py#L586

The error is silent.

A workaround is to split the mean along two or more axes.

A stopgap fix would be to throw an error if the tensor is too large, so that the user can decide how to deal with it.

Another fix would be to sum in int64.

**Comment by jakevdp** [comment]
Thanks for the report – I vaguely recall this coming up recently in the context of `mean` as well.

Perhaps since we end up converting to float in the end anyway, a better fix would be to accumulate the normalizer in float to avoid this kind of overflow? I'm trying to think whether there would be any downsides to that.
